Move wiki to developer guide in docs

dmentipl opened the pull request:
Addresses #1451

This PR adds content from the [IMPROVER wiki](https://github.com/metoppv/improver/wiki) to a "developer guide" section in the docs. Quoting from issue #1451:

> The content of the wiki is useful for developers. Moving it under source control means that changes to it require review, thus checking that the content is relevant and correct.

I have added all pages from the wiki to `docs/source` and changed the format from Markdown to reStructuredText, as expected by Sphinx, and fixed some minor formatting issues. I used [pandoc](https://pandoc.org/) to do this automatically after which I checked over the output manually. I have added a new "toctree" for the new content to `index.rst` under the heading "Developer Guide".

If this PR is accepted, I expect that we will remove the wiki.

Testing:
 - [x] Ran tests and they passed OK

bayliffe commented on doc/source/Ticket-Creation-and-Definition-of-Ready.rst:
```diff
@@ -0,0 +1,103 @@
+on the tickets for discussion when the ticket is started. \* Use user
+story format “As … I want … so that …”. \* Be clear enough so there is a
+shared understanding of the ticket. \* List all dependencies and related
+work on the ticket.
```
The bulleted lists in this file were a mix of hyphen and asterisk definitions; no reason for the inconsistency. The automatic translation has escaped all the asterisks, meaning these are no longer bulleted lists:
https://bens-improver-fork.readthedocs.io/en/wikitest/Ticket-Creation-and-Definition-of-Ready.html

dmentipl replied:
I've made unordered lists consistent (using `*`).

bayliffe commented on doc/source/Ticket-Creation-and-Definition-of-Ready.rst:
```diff
@@ -0,0 +1,103 @@
+----------------
+
+A number of labels are available for helping the categorise the
+available issues. Existing labels are: ### Type \* Administrative: The
```
The Type heading is not encoded correctly.

dmentipl replied:
Fixed.

bayliffe commented on doc/source/Running-at-your-site.rst:
````diff
@@ -0,0 +1,155 @@
+Required dependencies are listed in
+```environment_py36_iris22.yml`` <https://github.com/metoppv/improver/blob/master/envs/environment_py36_iris22.yml>`__
+and
+```environment_py37_iris24.yml`` <https://github.com/metoppv/improver/blob/master/envs/environment_py37_iris24.yml>`__,
````
These display quite unpleasantly:
https://bens-improver-fork.readthedocs.io/en/wikitest/Running-at-your-site.html

dmentipl replied:
Unfortunately, it seems not possible (or least complicated) to get "monospace" font in a hyperlink. rST and Sphinx do have some quirks. I've made the links "regular" text with the hyperlink.

bayliffe commented on doc/source/Guidance-for-Reviewing-Code.rst:
```diff
@@ -0,0 +1,108 @@
+   6. **Ensure their testing is documented on the issue.**
+
+4. Main reviewers should post comments to the pull request to show that
+   they have completed: i, ii, iii, iv, v, vi.
```
The further indented bulleted list above has been relabelled with numbers rather than roman numerals, which means it no longer matches with this list of tasks to complete. Please can you use roman numerals (or letters) for the above else there is no way to distinguish between the top level numbered bullets and this inset list.

dmentipl replied:
I've fixed this. Plus it seems Sphinx requires very particular alignment.

bayliffe replied:
Apparently there's a known issue with read-the-docs inheriting the format of the parent list for sublists in some cases:
https://github.com/readthedocs/sphinx_rtd_theme/issues/484

As such this change still renders on read-the-docs as a numbered list. Using letters seems to work both locally and on read-the-docs, so I've pushed up a change to do that.

https://bens-improver-fork.readthedocs.io/en/wikitest/Guidance-for-Reviewing-Code.html

bayliffe replied:
I'm happy with your changes, and my alterations on top are tiny. As such, I'm going to merge this and close down the wiki Thanks @dmentipl.

dmentipl replied:
Thanks for looking into this and providing a solution 👍

bayliffe commented on doc/source/Guidance-for-Reviewing-Code.rst:
```diff
@@ -0,0 +1,108 @@
+
+      -  Do the tests provided cover the expected situations?
+      -  Are the tests understandable?
+      -  Have edge cases been considered?
```
Again, it would be nice if the inset list used different labels.

dmentipl replied:
Fixed.

bayliffe commented on doc/source/Guidance-for-Reviewing-Code.rst:
```diff
@@ -0,0 +1,108 @@
+
+Always rerun Travis before merging a PR. Do this by: 1. At the bottom of
+the PR, click ``show all checks`` and then ``details`` next to the
+Travis check. 2. Click ``Restart build``
```
1 and 2 here should be a bulleted list to make it clearer; it's become inline in the translation process.

dmentipl replied:
Fixed.

bayliffe commented on doc/source/Code-Style-Guide.rst:
```diff
@@ -0,0 +1,746 @@
+suggests that for reviews to be effective reviewers should: \* Take time
+over reviews: don’t review faster than 500 lines of code per hour \* Not
+review more than 400 lines of code at a time, or for more than one hour
+continuously
```
Another bulleted list that has been inlined.

dmentipl replied:
Fixed.

bayliffe commented on doc/source/Code-Style-Guide.rst:
```diff
@@ -0,0 +1,746 @@
+Compulsory arguments. \* Keyword Args : Keyword arguments. \* Raises:
+Exceptions raised. \* Returns: Variables returned by the function or
+method. \* References: Link to available documentation. \* Warns:
+Warnings raised.
```
Another bulleted list that has been inlined.

dmentipl replied:
Fixed.

bayliffe commented on doc/source/Code-Style-Guide.rst:
```diff
@@ -0,0 +1,746 @@
+runtime, hence the need to define these clearly within the docstring.
+
+Within docstrings, when specifying a variable type, Python built-in data
+types can be used directly e.g. \* int \* float \* str \* bool
```
Another bulleted list that has been inlined.

dmentipl replied:
Fixed.

bayliffe commented on doc/source/Code-Style-Guide.rst:
```diff
@@ -0,0 +1,746 @@
+the ``intersphinx_mapping`` needs to be updated to link to the
+documentation of the module where the variable type originates. For
+example: \* numpy.ndarray \* datatime.datetime The full name of the
+module is required, rather than an alias e.g. numpy rather than np.
```
Another bulleted list that has been inlined.

dmentipl replied:
Fixed.

bayliffe commented on doc/source/Code-Style-Guide.rst:
```diff
@@ -0,0 +1,746 @@
+Helper functions or methods may live in one of several places. Where
+they should live depends on how they will be used.
+
+**Case 1:** A function used in more than one module. - In this case the
+function should be located in a shared location e.g. utilities.py
+
+**Case 2:** A function used by several classes within one module. - The
+function should be kept outside of any one class, but within the module.
+
+**Case 3:** A function used only within a single class. - Should be kept
+within the class; as a static method if it makes no use of self.
```
More bullets that has been inlined.
https://bens-improver-fork.readthedocs.io/en/wikitest/Code-Style-Guide.html#helper-functions-methods

dmentipl replied:
Fixed.

bayliffe commented on doc/source/Code-Style-Guide.rst:
```diff
@@ -0,0 +1,746 @@
+given which is a path to a cube to load - ``cli.inputjson`` - Where a
+string is given which is a path to a json file to load -
+``cli.comma_separated_list`` - This will convert the argument into that
+format and deal with error handling if no conversion is possible.
```
Another bulleted list that has been inlined.

dmentipl replied:
Fixed.

bayliffe commented on doc/source/Code-Style-Guide.rst:
```diff
@@ -0,0 +1,746 @@
+               coord_for_masking,
+               radius: float = None)
+
+The required arguments in this example are: - cube - coord_for_masking
```
Another bulleted list that has been inlined.

dmentipl replied:
Fixed.

bayliffe commented on doc/source/Code-Style-Guide.rst:
```diff
@@ -0,0 +1,746 @@
+
+You should add a unit test if: \* the logic in the function or method is
+not reasonably obvious from just looking at it, or \* the code is not
+otherwise covered by unit tests
```
Another bulleted list that has been inlined.

dmentipl replied:
Fixed.

bayliffe commented on doc/source/Code-Style-Guide.rst:
```diff
@@ -0,0 +1,746 @@
+ensure these are represented in the unit tests. 4. Consider possible
+edge cases e.g. cubes with different input dimensions. 5. Ensure the
+correct output is generated in good cases. 6. Ensure exceptions are
+raised as required for bad cases.
```
Another bulleted list that has been inlined.

dmentipl replied:
Fixed.

bayliffe commented on doc/source/Code-Style-Guide.rst:
```diff
@@ -0,0 +1,746 @@
+
+New unit test in IMPROVER should be written adhering to pytest style.
+These include: \* no test classes \* tests are written as functions \*
+test data is provided by fixtures
```
Another bulleted list that has been inlined.

dmentipl replied:
Fixed.

bayliffe reviewed (changes requested):
Thanks for this Daniel, it will look a lot nicer. Mostly bulleted lists that need sorting to make them render correctly.

dmentipl commented:
Thanks for the review @bayliffe. Sorry, I missed quite a bit of formatting issues caused by the pandoc conversion. Thanks for picking them up. I have fixed all the points you've raised, and corrected a few other issues I discovered. Hopefully it's ready to go now.
